llvm/lib/LTO/ThinLTOCodeGenerator.cpp

```
@@ -439,6 +452,259 @@ class ModuleCacheEntry {
+
+    ID = *Key;
+  }
+
+  std::string getEntryPath() final { return ID; }
+
+  // Try loading the buffer for this cache entry.
+  ErrorOr<std::unique_ptr<MemoryBuffer>> tryLoadingBuffer() final {
+    if (ID.empty())
+      return std::error_code();
+
+    // Lookup the output value from KVDB.
+    auto GetResponse = Service.KVDB->getValueSync(ID);
+    if (!GetResponse)
+      return errorToErrorCode(GetResponse.takeError());
+
+    // Cache Miss.
+    if (!*GetResponse)
+      return std::error_code();
+
+    // Malformed output. Error.
+    auto Result = (*GetResponse)->find("Output");
+    if (Result == (*GetResponse)->end())
+      return std::make_error_code(std::errc::message_size);
+
+    if (DeterministicCheck)
+      PresumedOutput = Result->getValue();
+
+    // Request the output buffer.
+    auto LoadResponse = Service.CASDB->loadSync(Result->getValue(), OutputPath);
+    if (!LoadResponse)
+      return errorToErrorCode(LoadResponse.takeError());
+
+    // Object not found. Treat it as a miss.
+    if (LoadResponse->KeyNotFound)
+      return std::error_code();
+
+    ProducedOutput = true;
+
+    SmallString<64> ResultPath;
+    Expected<sys::fs::file_t> FDOrErr = sys::fs::openNativeFileForRead(
+        Twine(OutputPath), sys::fs::OF_UpdateAtime, &ResultPath);
+    if (!FDOrErr)
+      return errorToErrorCode(FDOrErr.takeError());
+    ErrorOr<std::unique_ptr<MemoryBuffer>> MBOrErr = MemoryBuffer::getOpenFile(
+        *FDOrErr, OutputPath, /*FileSize=*/-1, /*RequiresNullTerminator=*/false);
+    sys::fs::closeFile(*FDOrErr);
+    return MBOrErr;
+  }
+
+  // Cache the Produced object file
+  void write(const MemoryBuffer &OutputBuffer) final {
+    if (ID.empty())
+      return;
+
+    if (!ProducedOutput)
+      cantFail(ModuleCacheEntry::writeObject(OutputBuffer, OutputPath));
+
+    ProducedOutput = true;
+    auto SaveResponse = Service.CASDB->saveFileSync(OutputPath);
+    if (!SaveResponse) {
+      consumeError(SaveResponse.takeError());
```

I'd prefer not to hide error responses, is it possible to report a failure if this occurs? (same comment for the other places that do `consumeError`).

---

Just print something to stderr or error out? While none of those are expected to fail but the recovering from thinLTO pipeline is super easy and user will not be hit with error when service is not available.

---

I'd like to error out. If I expect to be using the service during building, but the service is have trouble for some reason, I'd like to get the failure as an error.

Whether the service wants to provide an error or not it's up to the service (e.g. maybe the service chooses to ignore network errors), but if the service provides an error we should fail the build and report it.